### PR TITLE
Search engine result pages are not being classified

### PR DESCRIPTION
**Pull request by tmancey**
Fixes https://github.com/brave/brave-browser/issues/6000

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Confirm SERP for the following are classified:

https://amazon.com
https://baidu.com
https://bing.com
https://duckduckgo.com
https://fireball.com
https://github.com
https://google.com
https://google.ca
https://google.co.uk
https://google.co.jp
https://stackoverflow.com
https://developer.mozilla.org
https://twitter.com
https://en.wikipedia.org
https://search.yahoo.com
https://search.yahoo.co.jp
https://youtube.com
https://startpage.com
https://infogalactic.com
https://wolframalpha.com
https://semanticscholar.org
https://qwant.com
https://yandex.com
https://ecosia.org
https://searx.me
https://findx.com

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 1]
```diff
@@ -968,7 +968,9 @@ std::vector<AdInfo> AdsImpl::GetEligibleAds(
     const std::vector<AdInfo>& ads) {
   std::vector<AdInfo> eligible_ads = {};
 
-  for (const auto& ad : ads) {
+  auto unseen_ads = GetUnseedAdsAndRoundRobinIfNeeded(ads);
```
Comment (emerick) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: Typo in method name (should be `Unseen`)?

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: thanks sir :-), all fixed

File: vendor/bat-native-ads/src/bat/ads/internal/search_providers.cc [hunk 2]
```diff
@@ -45,4 +66,18 @@ bool SearchProviders::IsSearchEngine(const std::string& url) {
...
 }
 
+bool SearchProviders::HasResultsPagesForUrl(
+    const std::string& url,
+    const SearchProviderInfo& search_provider) {
+  size_t index = search_provider.search_template.find('{');
+  std::string substring = search_provider.search_template.substr(0, index);
+  size_t href_index = url.find(substring);
+
+  if (index == std::string::npos || href_index == std::string::npos) {
```
Comment (emerick) [inline vendor/bat-native-ads/src/bat/ads/internal/search_providers.cc]: We're checking down here if index is an invalid value, but we've already used it up above. I think it'd be better to just split this up into two individual if statements and test them as soon as they're set.

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/search_providers.cc]: fixed, thanks

File: vendor/bat-native-ads/src/bat/ads/internal/search_engine/search_providers.h [hunk 3]
```diff
@@ -49,12 +49,22 @@ const std::vector<SearchProviderInfo> _search_providers = {
       // TODO(https://github.com/brave/brave-browser/issues/8487): Brave Ads
       // search providers definition doesn't match all patterns
       "https://google.com",
-      "https://www.google.com/search?q={searchTerms}",
+      "https://www.google.com/search?{searchTerms}",
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/search_engine/search_providers.h:52]: don't we need to include the `q` as key here here so `GetValueForKeyInQuery` knows which value to retrieve? Maybe we can change `SearchProviderInfo` to include a `key` property, then we don't need to do the regex bit on the pattern.

**Comment by tmancey** [comment]
Closing PR as stale, will open a new PR when work is prioritized in the future

**Pull request closed without merging**